t/unit/backends/test_couchdb.py

```diff
@@ -30,26 +30,77 @@ def test_init_no_pycouchdb(self):
             module.pycouchdb = prev
 
     def test_get_container_exists(self):
-            self.backend._connection = sentinel._connection
-            connection = self.backend.connection
-            assert connection is sentinel._connection
-            self.Server.assert_not_called()
+        self.backend._connection = sentinel._connection
+        connection = self.backend.connection
+        assert connection is sentinel._connection
+        self.Server.assert_not_called()
 
     def test_get(self):
         """test_get
 
         CouchBackend.get should return  and take two params
         db conn to couchdb is mocked.
-        TODO Should test on key not exists
-
         """
         x = CouchBackend(app=self.app)
         x._connection = Mock()
         get = x._connection.get = MagicMock()
-        # should return None
         assert x.get('1f3fab') == get.return_value['value']
         x._connection.get.assert_called_once_with('1f3fab')
 
+    def test_get_non_existent_key(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+        get = x._connection.get = MagicMock()
+        get.side_effect = pycouchdb.exceptions.NotFound
+        assert x.get('1f3fab') is None
+        x._connection.get.assert_called_once_with('1f3fab')
+
+    def test_set(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+
+        x.set('1f3fab', 'value')
+
+        x._connection.save.assert_called_once_with({'_id': '1f3fab',
+                                                    'value': 'value'})
+
+    def test_set_with_conflict(self):
+        x = CouchBackend(app=self.app)
+        x._connection = Mock()
+        x._connection.save.side_effect = (pycouchdb.exceptions.Conflict, None)
+        get = x._connection.get = MagicMock()
+
+        x.set('1f3fab', 'value')
+
+        x._connection.get.assert_called_once_with('1f3fab')
+        x._connection.get('1f3fab').__setitem__.assert_called_once_with(
```

Comment:
@thedrow I found this part a bit obscure. Just a suggestion then:

```
key = '1f3fab'
already_stored_data = {'_id': key, 'value': 'old_value'}
get = x._connection.get = MagicMock(return_value=already_stored_data)
updated_data = {'_id': key, 'value': 'new_value'}
x.set(key, updated_data['value'])

x._connection.get.assert_called_once_with(key)
# Dictionary should be mutated now, but I don't think we need this, because of the next assertion
assert get.return_value == updated_data
x._connection.save.assert_called_with(updated_data)
```

Reply:
I think I'll keep this test as it is. There's no reason to change a test that already proves that the code works.
Feel free to issue a new PR about it if you feel strongly about it :)